### DISCUSSION TITLE
Overhaul

### DISCUSSION THREAD

**Discussion by taux1c** [DIGITALCRIMINALS/OnlyFans#170] [Ideas]
There is no notation in the code, the readme is outdated, there are so many patches that if you don't read most of the closed issues you can't even get this to work and I can't even get it to work without running it in docker and after this x-bc issue I can't run it at all. I was going to try to help fix it but the code is a mess. 

Am I the only one that feels this way or are there others that would like the help me with getting it up to standard?

**Comment by stranger-danger-zamu** [comment]
Oh boy, here we go:

> I started modernizing the project minutiae (eg. pyproject.toml over requirements.txt) to get it to the point where it could be packaged into a zipapp using shiv. This would take care of installing the dependencies for all users that aren't messing with the codebase along with making it super simple for less technical users (ie. "Install python, download this, and just run this command and edit the files it tells you to"). No messing with virtualenvs, or clobbering system packages.
>
> It's horrify though. There is so much use of traversing the *installation* directory to do basic things.
>
> Not to mention that the configuration files are hardcoded internal to the *installation* directory. I've already sorted it out to use the `appdirs` module, but there probably needs to be some backwards compatible code to make it not break everything for the technically inept.
> 
> Also the naming. The project has a number of internal naming issues. The "tests" in the test directory are actually validations and assertions to prevent people from running things wrong. Additionally, the projects name is OF, which *__is__* most likely a trademark and something that they can go after this for so... lol. 
>
> Nearly all the user visible code is also sprinkled throughout the code base instead of delineating the function from the CLI. 
> 
> Also, the fact that the interactive session requires you to re-auth is ridiculous. It's just looping in the wrong place!
> 
> That said, I don't care enough to really rewrite the code base and look into streamlining some stuff. So far I'm just going through and changing the imports, moving *code* into a separate directory, and moving the configuration out of the installation directory. With that I can at least not care how much of a mess the code base is since it'll be neatly packaged away. 
> 
> I might get around to re-doing the CLI though since the UI/UX is a bit of a trash fire.

tl;dr: kinda started doing that already.

---

NB: Oh I haven't pushed or made any noise about it, since it's much easier to say "Hey change your entire workflow and code organization because I said so" when you are actually improving it and making it easier on them.

**Reply by stranger-danger-zamu** [reply]
If you were going to actually work on it though I definitely recommend taking a look at the use cases needed and abstracting behavior needed in the code base and making some abstract base classes. At least some interfaces (err, "Protocols" in python land) so the core logic can be written without caring for the target site. 

Also using something like pydantic for defining and validating configs. So you can tell the user what is wrong in their config other than just screeching that something is wrong. (You could even have program capture the errors and provide a link to how to fix them in documentation, but lol).

**Reply by stranger-danger-zamu** [reply]
Oh and replace all of os.path uses with pathlib.Path

**Reply by DIGITALCRIMINAL** [reply]
Make a commit lmao

**Reply by stranger-danger-zamu** [reply]
[k](https://github.com/DIGITALCRIMINALS/OnlyFans/issues/343)

**Comment by taux1c** [comment]
Yeah I figured someone had already started on this. Honestly some of it is over my head sadly. Given what I know though I see so many things that need to be changed but every time I look at if it could be done I see other things that tie in and I'm just like if this is taken out we would have to patch it in so many places but it needs to be moved. 

So my thought was if there was a group of us we could just simply have a second repo and start from scratch with a checks directory that would contain functions for checking where we just simply use a decorator to run checks. 

Then there is a browser extension called onlyfans cookie helper that is made for this program. It simply generates the auth.json. We could add a repo for that and plan for it accordingly. (It isn't maintained I don't think.) 

Basically I thought standards would be good because we have checks in a folder and then the program could be an engine that simply loads a config based on what site it is downloading from. (onlyfans, fansly, just4.fans, avn stars, etc.) Then we just make a new config file and it works with a new site. I understand it may not be a simple define variables config file but more like a module. like onlyfans module says send me all stuff labled <video> from onlyfans.com using this authentication file. the core would simply return the requested info and the module would compile a list and say okay go get all of this and return it. core returns it to the module and module handles saving per config for the site.

This is tons of parts thrown in one box and it needs to be put in boxes and labeled.

**Comment by DIGITALCRIMINAL** [comment]
@taux1c 
>No notion

Yeah, even I don't understand what's happening with the script and I'm trying to make it more readable by adding in Python types, and notes so whoever is forking it can easily add to it, but there's a lot to change.

>Readme is outdated

If you're talking about the readme.md, yeah probably. I forget to update it.

>I can't even get it to work without running it in docker

I assume you're on Linux, there was only 3 things I had to do to get the script running on Arch
Install Python 3.10, add in my auth.json and run the script. There was a 4th for Windows which required you to install Microsoft Build Tools, but that got removed.

I've overhauled the script multiple times, that's why there's so much legacy/dead code that need to be updated or deleted and people have been helping improve the script by adding commits.

**Reply by taux1c** [reply]
I'm so glad you're cool about it. I was skeptical about posting it because I wasn't trying to upset anyone. It's an awesome script. I think it just grew so fast that it didn't get split properly. I am gifted in that I see efficiency opportunity whenever I look at coding. 

I forget notes a lot and updating the readme also slips my mind almost regularly. That's why I use the to do tracker. I can add things like fix this, update readme, etc. 

I have so many ideas for this. Like I was thinking if we just made the core we could load a json config file as a module for new sites. Then the core could just say okay I need to get x elements and save them like this and this is how I sign in. Etc.

Then adding a new site is as simple as adding a new json. Just have a loop menu that lists all json directions in the menu folder that have a menu_name option or something.

**Reply by DIGITALCRIMINAL** [reply]
I don't really get what you're saying about creating a "json core", but the script relies on API calls to the websites it supports.

`Then the core could just say okay I need to get x elements and save them like this and this is how I sign in. Etc.`
The script already does this whole process from https://github.com/DIGITALCRIMINALS/OnlyFans/blob/ed356dc834d12f5795a6f6f997e2b408480fddab/datascraper/main_datascraper.py#L42

**Comment by hackerin0** [comment]
Oh I saw a lot of potential for this and since a month or so I am working on a new and improved version, which I am still thinking if I am going to release as open source.
I made a nice Tkinter gui, I am finishing up the styles, and and automated cookie catching for all the data needed, with also chance to evade captchas.
My idea is to make this A LOT easier, and make it usable also for people who hasn't an IT background. DIGITALCRIMINAL's work is really great, but it can be troublesome to get the whole thing working even for me, a web developer, I can't imagine what could the random chad do! 
My idea is to package everything needed in a nice little exe, it will show a prompt for username and password and then grab cookies with headless browser. Now I am trying to improve the whole interface thing, to be able to choose girls with checkboxes and stuff. 
I hope my release will be much, much easier to use (the idea is to make a click&download app!) and release it everywhere I can. The more we simplify and share, the more leaks will come :D 


Yo ho!

**Reply by DIGITALCRIMINAL** [reply]
We need more noob friendly scrapers :sob: 

If you release it on Github, I will link it in the docs :+1:

**Reply by hackerin0** [reply]
I am not totally sure about releasing on Github. Look at the success of this: main pirate porn sites and forum are linking to YOUR tool! You may have underestimated the greatness of what you did here! Models complaining about leaks of their whole stuff and shutting down accounts, and most important OF changing its site and mechanisms to counter-act to what we are doing here!

I think OF employees being able to just come here, check the source and the solutions you so hardly developed to scrape their site, and in 10 minutes being able to search your code, find what you do and make some counter-measures to it is NOT a great idea.
I am a Free software enthusiast, a RMS fan,  BUT for THIS particular script!..... I think it could be best for all the fapper pirates out there to keep the source secret or at least private between the people involved in its development.

Keeping it here will ALWAYS result in OF putting countermeasures in place, they can just search google for "how to scrape OF" and soon or late this repo will come out. They read your code, patch theirs and here we go again, with a bunch of noobs crying in the issues because they can't connect and with you brainfucking yourself to understand what those rich fuckers at OF are doing to lock us out from the stuff WE PAID FOR.

I also fear closed source software, especially pirate illegal closed source software that serves scraping a "porn" site, but people nowadays are just downloading shit here and there. "chaturbate token generators" are still a thing. You / we should do the jump, and close this thing, make a website and host your / "our" stuff there. 
<img src="https://user-images.githubusercontent.com/83502965/147395498-c458e892-22a3-4b83-b536-a08a90d8cab2.jpg" width="300"/>
And just release a super-simple noob-proof executable to spoil those bitches :D

Edit: Spelling

**Reply by taux1c** [reply]
My core thought was basically make the scraper full of variables that would be filled by json info almost like kwargs...
base_site = onlyfans.com/login...

And eventually configurations would be loaded by the scraper telling it what and how to scraped for that site.

Kind of like a scraper with its own api.

**Reply by taux1c** [reply]
Also, I agree keeping o.f. devs from the source code is kind of essential as much as I hate to say it closing the source may be the best move.

I wanted to turn it into a django or flask app where users could sign in or just hit go and grab the cookie. Then rip all the stuff and make a downloadable then also save a copy for a leak site.

**Reply by taux1c** [reply]
The leak site could actually be django or flask and load the content right in instantly.

**Reply by stranger-danger-zamu** [reply]
Closing source will just mean that there will be less people grabbing the content from their accounts and ergo less content being released in aggregate. It's a bad move overall.

**Reply by OFfriend** [reply]
We don't need closed source. The scraper simply does the same like the OF website. The only changing part is the signing parameters.

**Comment by taux1c** [comment]
I love the idea, I'm not sure how much I would use it. I had this running in a docker container on a headless Ubuntu server in my garage with a cron job to start it every hour.

Then using samba I had them on my desktop, phone, laptop, etc.
